core/include/traccc/edm/details/container_element.hpp

```diff
@@ -21,11 +21,17 @@ namespace traccc {
  * low-overhead access to the struct-of-arrays container class to emulate an
  * array-of-structs architecture.
  *
- * @tparam header_t The type of the header object.
- * @tparam vector_t The fully qualified vector type.
+ * @tparam header_reference_t The type of reference of the header object.
+ * @tparam vector_reference_t The fully qualified vector reference type.
  */
-template <typename header_t, typename vector_t>
+template <typename header_reference_t, typename vector_reference_t>
 struct container_element {
+
+    /// Header reference type
+    using header_reference = header_reference_t;
+    /// Vector reference type
+    using vector_reference = vector_reference_t;
+
```

Comment:
Please add a static assertion to assert that these types are either primitives, references, or pointers, because this is a massive amount of (performance) bug surface area.

Reply:
But they aren't. This is my whole point. When you access an "inner vector" in a `vecmem::jagged_device_vector`, you access a `vecmem::device_vector` that was created on the fly on the stack. Don't you remember?

The "device version" of this will have `header_reference == header_t&` and `vector_reference == vecmem::device_vector<item_t>`. By (current) design...

Reply:
Ah, I see. I haven't used jagged vectors much, so I was not aware. Still, is there a reason why we want these types to be so general? I don't necessarily see why a more specific template type (from which we can then construct the device vector types) wouldn't suffice?